Adopt more spans in CSSParserTokenRange

cdumez opened the pull request:
#### 643a222312ff1d998bd15e8fedf5a7d7441b2cc8
<pre>
Adopt more spans in CSSParserTokenRange
<a href="https://bugs.webkit.org/show_bug.cgi?id=286214">https://bugs.webkit.org/show_bug.cgi?id=286214</a>

Reviewed by Geoffrey Garen.

This tested as performance neutral on Speedometer.

* Source/WebCore/css/parser/CSSParserTokenRange.cpp:
(WebCore::CSSParserTokenRange::makeSubRange const):
(WebCore::CSSParserTokenRange::consumeBlock):
(WebCore::CSSParserTokenRange::consumeBlockCheckingForEditability):
(WebCore::CSSParserTokenRange::consumeComponentValue):
(WebCore::CSSParserTokenRange::trimTrailingWhitespace):
(WebCore::CSSParserTokenRange::consumeLast):
(WebCore::CSSParserTokenRange::serialize const):
* Source/WebCore/css/parser/CSSParserTokenRange.h:
(WebCore::CSSParserTokenRange::CSSParserTokenRange):
(WebCore::CSSParserTokenRange::atEnd const):
(WebCore::CSSParserTokenRange::begin const):
(WebCore::CSSParserTokenRange::end const):
(WebCore::CSSParserTokenRange::size const):
(WebCore::CSSParserTokenRange::peek const):
(WebCore::CSSParserTokenRange::consume):
(WebCore::CSSParserTokenRange::consumeIncludingWhitespace):
(WebCore::CSSParserTokenRange::consumeWhitespace):
(WebCore::CSSParserTokenRange::consumeAll):
(WebCore::CSSParserTokenRange::span const):
(): Deleted.

Canonical link: <a href="https://commits.webkit.org/289131@main">https://commits.webkit.org/289131@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d411aa0dab758bc0dabe4201dc15bd2079837720

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85415 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5141 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39838 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90534 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36448 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87495 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5260 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13119 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66398 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24211 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88453 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4019 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77572 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46680 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3903 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31833 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35518 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/78401 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74605 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32680 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92071 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/84447 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12754 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9340 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75019 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12982 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/73405 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74140 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18477 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16907 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4800 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13329 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12722 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18166 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/106837 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12545 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25746 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16019 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14304 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebCore/css/parser/CSSParserTokenRange.cpp:
```diff
@@ -103,31 +112,31 @@ void CSSParserTokenRange::consumeComponentValue()
-    }
+    size_t i = m_tokens.size();
+    for (; i > 0 && CSSTokenizer::isWhitespace(m_tokens[i - 1].type()); --i) { }
+    m_tokens = m_tokens.first(i);
```
This could be

```
while (m_tokens.size() && CSSTokenizer::isWhitespace(m_tokens.last())) {
    dropLast(m_tokens)
}
```

(if you added a WTF::dropLast helper function)

cdumez replied:
Hopefully this is identical perf-wise since I A/B tested this. I was trying to update the span only once instead of once per loop iteration but maybe it doesn't matter.

geoffreygaren replied:
I suspect it won't matter because the previous code was happy to update m_last, which is pretty much identical.

geoffreygaren commented on Source/WebCore/css/parser/CSSParserTokenRange.cpp:
```diff
@@ -103,31 +112,31 @@ void CSSParserTokenRange::consumeComponentValue()
         eofToken();
-    return *--m_last;
+    auto* last = &m_tokens.back();
+    m_tokens = m_tokens.first(m_tokens.size() - 1);
```
Could also use WTF::dropLast here (or maybe a WTF::consumeLast).

geoffreygaren commented on Source/WebCore/css/parser/CSSParserTokenRange.h:
```diff
@@ -88,34 +89,30 @@ WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
-            ++m_first;
+        size_t i = 0;
+        for (; i < m_tokens.size() && CSSTokenizer::isWhitespace(m_tokens[i].type()); ++i) { }
+        m_tokens = m_tokens.subspan(i);
```
Could be

```
while (m_tokens.size() && CSSTokenizer::isWhitespace(m_tokens.first())) {
    skip(m_tokens, 1)
}
```

cdumez committed:
Adopt more spans in CSSParserTokenRange
https://bugs.webkit.org/show_bug.cgi?id=286214

Reviewed by Geoffrey Garen.

This tested as performance neutral on Speedometer.

* Source/WebCore/css/parser/CSSParserTokenRange.cpp:
(WebCore::CSSParserTokenRange::makeSubRange const):
(WebCore::CSSParserTokenRange::consumeBlock):
(WebCore::CSSParserTokenRange::consumeBlockCheckingForEditability):
(WebCore::CSSParserTokenRange::consumeComponentValue):
(WebCore::CSSParserTokenRange::trimTrailingWhitespace):
(WebCore::CSSParserTokenRange::consumeLast):
(WebCore::CSSParserTokenRange::serialize const):
* Source/WebCore/css/parser/CSSParserTokenRange.h:
(WebCore::CSSParserTokenRange::CSSParserTokenRange):
(WebCore::CSSParserTokenRange::atEnd const):
(WebCore::CSSParserTokenRange::begin const):
(WebCore::CSSParserTokenRange::end const):
(WebCore::CSSParserTokenRange::size const):
(WebCore::CSSParserTokenRange::peek const):
(WebCore::CSSParserTokenRange::consume):
(WebCore::CSSParserTokenRange::consumeIncludingWhitespace):
(WebCore::CSSParserTokenRange::consumeWhitespace):
(WebCore::CSSParserTokenRange::consumeAll):
(WebCore::CSSParserTokenRange::span const):
(): Deleted.

Canonical link: https://commits.webkit.org/289131@main
